### PR TITLE
172911 - decrypt permit keys to extract enc keys

### PR DESCRIPTION
**Pull request by MastekRohitChettri**
This PBI consists changes of [PBI : 172911](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/172911)

- Code changes added
- Unit test added

### CODE REVIEW

File: src/UKHO.S100PermitService.Common/Encryption/AesEncryption.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Security.Cryptography;
+
+namespace UKHO.S100PermitService.Common.Encryption
+{
+    [ExcludeFromCodeCoverage]
+    public class AesEncryption : IAesEncryption
+    {
+        private const int KeySize = 128;
```
Comment (harshal11869) [inline src/UKHO.S100PermitService.Common/Encryption/AesEncryption.cs:9]:
```suggestion
        private const int KeySize = 128, BlockSize = 128;
```

File: src/UKHO.S100PermitService.Common/Encryption/AesEncryption.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            var aes = Aes.Create();
+            aes.BlockSize = KeySize;
+            aes.KeySize = KeySize;
+            aes.IV = new byte[IvLength];
+            aes.Mode = CipherMode.CBC;
+            aes.Padding = PaddingMode.None;
+            aes.Key = StringToByteArray(keyHexEncoded);
+            return aes;
+        }
```
Comment (harshal11869) [inline src/UKHO.S100PermitService.Common/Encryption/AesEncryption.cs:33]:
```suggestion
        private static Aes CreateAes(string keyHexEncoded)
        {
            var aes = Aes.Create();
            aes.BlockSize = BlockSize;
            aes.KeySize = KeySize;
            aes.IV = new byte[IvLength];
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.Zeros;
            aes.Key = StringToByteArray(keyHexEncoded);
            return aes;
        }
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: src/UKHO.S100PermitService.Common/Encryption/AesEncryption.cs (hunk 1 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.Common/Encryption/AesEncryption.cs]: Resolved

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by amogh14893 — CHANGES_REQUESTED** [review]

File: Deployment/templates/continuous-deployment.yml
[the remark this answers is not in the document]
Comment (nerellavinod) [reply Deployment/templates/continuous-deployment.yml:1]: resolved comment

**Review by Nadeem14456 — APPROVED** [review]
LGTM.

File: tests/UKHO.S100PermitService.Common.UnitTests/Encryption/AesEncryptionTests.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        public void WhenProvidedValidData_ThenSuccessfullyReturnsDecryptedData()
+        {
+            var result = _aesEncryption.Decrypt(FakeText, FakeKey);
+
+            result.Should().NotBeNullOrEmpty();
+            result.Should().NotBe(FakeText);
+        }
+
+        [Test]
+        public void WhenProvidedValidData_ThenSuccessfullyReturnsEncryptedData()
```
Comment (jrippington) [inline tests/UKHO.S100PermitService.Common.UnitTests/Encryption/AesEncryptionTests.cs:29]: Test what happens if null, empty or incorrect length strings are passed in to the Encrypt()/Decrypt() methods?

**Review by jrippington — CHANGES_REQUESTED** [review]
Few extra tests needed

File: tests/UKHO.S100PermitService.Common.UnitTests/Encryption/AesEncryptionTests.cs (hunk 3 above)
Comment (MastekRohitChettri) [reply tests/UKHO.S100PermitService.Common.UnitTests/Encryption/AesEncryptionTests.cs:29]: Those cases are handled in **S100CryptTests** class.

Comment (jrippington) [reply tests/UKHO.S100PermitService.Common.UnitTests/Encryption/AesEncryptionTests.cs:29]: AesEncryption is a public class, so the tests for these conditions should be factored out

Comment (MastekRohitChettri) [reply tests/UKHO.S100PermitService.Common.UnitTests/Encryption/AesEncryptionTests.cs]: Code refactored.

**Review by amogh14893 — APPROVED** [review]

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Pull request merged**